apps/server/src/api-data/excel/parsing/utils.ts

```diff
@@ -0,0 +1,169 @@
+ * @private exported for testing
+ */
+export function getCustomFieldData(
+  importMap: ImportMap,
+  existingCustomFields: CustomFields,
+): {
+  mergedCustomFields: CustomFields;
+  customFieldImportKeys: Record<keyof CustomFields, string>;
+} {
+  const mergedCustomFields: CustomFields = {};
+  /**
+   * A map of import keys to ontime keys
+   * Map<excel column name, ontime key>
+   */
+  const customFieldImportKeys: Record<string, CustomFieldKey> = {};
+
+  for (const ontimeLabel in importMap.custom) {
+    // if the label is not valid, we skip the import
+    if (!checkRegex.isAlphanumericWithSpace(ontimeLabel)) {
+      continue;
+    }
+
+    // generate a key for the custom field
+    const keyInCustomFields = customFieldLabelToKey(ontimeLabel);
+    // we lower case the excel key to make it easier to match
+    const columnNameInExcel = importMap.custom[ontimeLabel].toLowerCase();
+    const maybeExistingColour = existingCustomFields[keyInCustomFields]?.colour ?? '';
+
+    // 1. add the custom field to the merged custom fields
+    mergedCustomFields[keyInCustomFields] = {
+      type: 'text', // we currently only support text custom fields
+      colour: maybeExistingColour,
+      label: ontimeLabel,
+    };
+
+    // 2. add the column to the import keys
+    customFieldImportKeys[columnNameInExcel] = keyInCustomFields;
+  }
+  return { mergedCustomFields, customFieldImportKeys };
+}
+
+/**
+ * Utility function infers a boolean from a string value
+ */
+export function parseBooleanString(value: unknown): boolean {
+  if (typeof value === 'boolean') {
+    return value;
+  }
+
+  // falsy values would be nullish or empty string
+  if (!value || typeof value !== 'string') {
+    return false;
+  }
+  return value.toLowerCase() !== 'false';
+}
+
+type IndexMap = Record<keyof Omit<ImportMap, 'worksheet' | 'custom'>, MaybeNumber> &
+  Record<keyof Pick<ImportMap, 'custom'>, Record<number, string>>;
+
+export type SheetMetadata = Partial<
+  Record<keyof Omit<ImportMap, 'worksheet' | 'custom'>, { row: number; col: number }> &
+    Record<string, { row: number; col: number }>
+>;
+
+export function generateImportHandlers(importMap: ImportMap) {
+  const indexMap: IndexMap = {
+    title: null,
+    cue: null,
+    note: null,
+    colour: null,
+    flag: null,
+    skip: null,
+    countToEnd: null,
+    linkStart: null,
+    timeStart: null,
+    timeEnd: null,
+    duration: null,
+    timeWarning: null,
+    timeDanger: null,
+    endAction: null,
+    timerType: null,
+    entryId: null,
+    custom: {},
+  };
+
+  const sheetMetadata: SheetMetadata = {};
+
+  const handlers = {
+    [importMap.timeStart]: (row: number, col: number) => {
+      indexMap.timeStart = col;
+      sheetMetadata.timeStart = { row, col };
+    },
+    [importMap.linkStart]: (row: number, col: number) => {
+      indexMap.linkStart = col;
+      sheetMetadata.timeEnd = { row, col };
```

Comment:
should this be `linkStart`?

```suggestion
      sheetMetadata.linkStart = { row, col };
```